### PR TITLE
Make `RustString` an extern type to avoid `improper_ctypes` warnings

### PR DESCRIPTION
**Pull request by Zalathar**
Currently, any FFI function that uses `&RustString` needs to also add `#[ignore(improper_ctypes)]` to silence a warning.

The warning is not _completely_ bogus, because `RustString` contains `Vec<u8>` and therefore does not have a guaranteed layout. But we have no way of telling the lint that this doesn't matter, because the C++ code only uses that pointer opaquely and never relies on its underlying layout.

Ideally there would be some way to silence `improper_ctypes` at the type-definition site. But because there isn't, casting to and from a separate extern type is better than having to annotate every single use site.

### CODE REVIEW

**Comment by Zalathar** [comment]
I also want to rename `RustString` to something else, because it's a raw byte buffer rather than a UTF-8 sequence, but that would be too much scope creep for this PR.

**Comment by Zalathar** [comment]
See also #116963 for general dissatisfaction with `improper_ctypes`.

**Comment by Zalathar** [comment]
I'm not 100% confident that casting between normal references and extern type references like this is OK, but it seems reasonable?

**Comment by Zalathar** [comment]
See https://github.com/rust-lang/rust/pull/67077/commits/47e932b96e726d855f9885f073ef6b6b6bb78438 for why RustString is in `rustc_llvm`.

**Comment by Zalathar** [comment]
Rebased to remove another `ignore` added by #132514 [(diff)](https://github.com/rust-lang/rust/compare/a1dcb0221669561060105950e9c8a821c9390842..730d08819d958b806f0117da427d5805202fdfc6).

**Comment by Zalathar** [comment]
Rebased to resolve conflicts and remove more ignore/expect added by (my) other PRs.

File: compiler/rustc_llvm/src/lib.rs [hunk 1]
```diff
@@ -2,42 +2,81 @@
 #![allow(internal_features)]
 #![doc(html_root_url = "https://doc.rust-lang.org/nightly/nightly-rustc/")]
 #![doc(rust_logo)]
+#![feature(extern_types)]
 #![feature(rustdoc_internals)]
 #![warn(unreachable_pub)]
 // tidy-alphabetical-end
 
-// NOTE: This crate only exists to allow linking on mingw targets.
```
Comment (Zalathar) [inline compiler/rustc_llvm/src/lib.rs:9]: I'm not sure what this was originally referring to, but it's pretty clearly no longer true of the present crate.

**Review by cuviper — CHANGES_REQUESTED** [review]
Overall it looks like a nice improvement! I just have a couple quibbles.

**Comment by Zalathar** [comment]
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/b266f0c9519b464247a2635f3ed3c06bef255dc7..89d7efaf8f5ed8c1c15faea21824ae479656bdc1).

File: compiler/rustc_llvm/src/lib.rs [hunk 2]
```diff
@@ -2,42 +2,81 @@
...
+    bytes: RefCell<Vec<u8>>,
+}
+
+impl RustStringInner {
+    fn from_opaque(opaque: &RustString) -> &Self {
+        // SAFETY: A valid `&RustString` must have been created via deref of
+        // `RustStringInner`.
+        let ptr: *const RustString = ptr::from_ref(opaque);
+        let ptr: *const RustStringInner = ptr.cast();
+        unsafe { ptr.as_ref() }.unwrap()
```
[the remark this answers is not in the document]
Comment (Zalathar) [reply compiler/rustc_llvm/src/lib.rs]: Yeah, I must have been influenced by other code where the null case is relevant, but here we start with a reference so the unwrap doesn't add anything.

**Pull request merged**